Sources/FoundationXML/XMLParser.swift

```diff
@@ -471,7 +471,7 @@ open class XMLParser : NSObject {
     open var allowedExternalEntityURLs: Set<URL>?
     
 #if os(WASI)
-    private static var _currentParser: XMLParser?
+    private nonisolated(unsafe) static var _currentParser: XMLParser?
```

Comment:
This seems unsafe at first look, since any thread can call `setCurrentParser` and effectively traffic `XMLParser` between actors. Can you elaborate a bit more on why this is necessary?

Reply:
WASI preview1 ABI is always single-threaded, so memory access to the same storage should not happen simultaneously. So any data storage should be safe to access from any isolation context on WASI, but  6 lang mode requires explicit marking here.

Reply:
Perhaps the better question is what does actor isolation mean on WASI where everything is single threaded? This will allow this mutable state to pass between actor isolations on WASI, is that ok because we know it's single threaded? Or should this instead be some form of task-local storage to prevent it from traveling across actors?

Reply:
We don't need task local storage at all here because there is no suspension point during `XMLParser.parse` call. No suspension point means no Task switching, thus it's guaranteed that `XMLParser.setCurrentParser(self)` is not called twice before resetting the storage by `XMLParser.setCurrentParser(nil)`.
The only exception is the case where user provided `XMLParserDelegate` methods call `XMLParser.parse()`, but the reentrancy issue is not specific to `os(WASI)` branch.

Reply:
Thanks for the explanation. The reentrancy possibility isn't unique to WASI, but its ability to manipulate the underlying `nonisolated(unsafe)` variable across actor isolations is unique to WASI with this change (on other platforms, that variable is thread-specific - which should probably be marked as unavailable from async for this reason).

The example I'm thinking of is: someone calls `XMLParser.parse()`, in a delegate callout they create a new detached task on a different actor calling `XMLParser.parse()` again. At this point, we have two actors manipulating the same underlying `nonisolated(unsafe)` variable. Typically, that'd be a concurrency violation because two actors are mutating the same shared state. Since WASI is single-threaded, we know that this can't data race in practice, but it's still a violation in the Swift 6 concurrency design and I'm not sure how to rationalize that with the thread model of WASI at the moment (cc @hborla in case you have any insight here)

Reply:
> The example I'm thinking of is: someone calls XMLParser.parse(), in a delegate callout they create a new detached task on a different actor calling XMLParser.parse() again.

Having a closer look, I noticed the problem exists even in non-WASI code path since jobs associated with two different Tasks can be scheduled on the same thread, and in that case two different tasks can manipulate the same thread local variable.

So we had two issues in the existing code, 1. reentrancy issue, 2. sharing mutable storage across Tasks. I'll send a fix for them in another PR and exclude the commit from this PR. Does it make sense?

Reply:
@jmschonfeld Opened another PR to fix the issues https://github.com/apple/swift-corelibs-foundation/pull/5061